Rule: Headers attribute only refers to cells in the same table (a25f45) (#412)
act-rules/act-rules.github.io: _rules/table-headers-attribute-refer-to-data-cells-a25f45.md

---
id: a25f45
name: headers attribute only refers to cells in the same table element.
rule_type: atomic
description: |
  This rule checks that the headers attribute must only refer to cells in the same table element.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied

>     passed: further testing needed

Author: There was never a test for this. Test only checks for top level frontmatter keys. I think we should add a test for this.

>     passed: further testing needed

Author: Added a test - PR https://github.com/act-rules/act-rules.github.io/pull/691

input_aspects:
  - DOM Tree
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

This rule applies to `td` elements with a `headers` attribute, that are [included in the accessibility tree](#included-in-the-accessibility-tree).

## Expectation

> Each target element refers in its `headers` attribute to other `th` elements or cells with a [semantic role](#semantic-role) of `columnheader` or `rowheader` within the same `table` element.

Reviewer: I don't really think this expectation is accurate enough. What does "refers in its headers attribute" mean? What is a "cell", when is something in the "same table"? I've tried to write this out myself and came up with the following:

> The `header` attribute of each target element is [a set of space separated IDs]([set of space tokens](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens)), each of which is an ID of an element that:
> 
> 1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and
> 2. is a `cell` of the same [`table`](#https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element

## Assumptions

- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.

> - Tables using `headers` id associations are intended as data rather than [layout table](#layout-table).

Reviewer: I don't think you need to define layout table here. Linking out to something about layout tables in WCAG is enough. We may need that term for some other rule.

Author: Agreed with @WilcoFiers, to leave this as is for now.

```html
<table>
  <thead>	
    <tr>
      <td role='columnheader' id="header1">Projects</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td headers="header1">15%</td>
    </tr>
  </tbody>
</table>
```

#### Passed Example 4

A `table` with multiple `columnheader(s)`, where the `headers` attribute on the cell lists the `id(s)` of `th` within the same `table`.

```html
<table>
  <tr>
    <th colspan="2" id="header1">Projects</th>
    <th colspan="2" id="header2">Exams</th>
  </tr>
  <tr>
    <th id="e1" headers="header1">1</th>
    <th id="e2" headers="header1">2</th>
    <th id="p1" headers="header2">1</th>
    <th id="p2" headers="header2">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 e2">15%</td>
    <td headers="header2 p1">15%</td>
    <td headers="header2 p2">45%</td>
  </tr>
</table>
```

### Failed

#### Failed Example 1

Cells with `headers` attribute, refers to non-existing `th` within the same `table`.

```html
<table>
  <tr>
    <th id="header1">Projects</th>
    <th id="header2">Exams</th>
  </tr>
  <tr>
    <td headers="header3">15%</td>
    <td headers="header3">15%</td>
  </tr>
</table>

> ```

Reviewer 2: I think this example would also fail an HTML validation test. Is this still something to include in an accessibility test? (just double checking)

Author: @EmmaJP in this case we are simply using a `header3` as an `id` that does not exist. I ran the HTML validation [`html-hint`](https://www.npmjs.com/package/htmlhint) against the above snippet and it seemed ok.

Reviewer 2: W3 nu validator throws an error when 'header 3' does not exist, and also when 'header 3' exists as an id in a different table. So this example is also invalid HTML.

I've always thought valid HTML a prerequisite for accessibility rules.

Author: Right. I think we will need to fix this out of scope of this PR for 2 reasons.

- I currently have a different pr, which checks validity of HTML using `html-hint`, see https://github.com/act-rules/act-rules.github.io/pull/631
- I would like to integrate something like w3 nu validator as a part of continuous integration testing.

In this specific example we have refer to an id that does not exist, so yes it will refer to an `does not exist` error.

I have created an issue for this Emma - https://github.com/act-rules/act-rules.github.io/issues/679

#### Failed Example 2

One of the cells with `headers` attribute, refers to `th` that does not exist within the same `table`.

```html
<table>
  <tr>
    <th colspan="2" id="header1">Projects</th>
  </tr>
  <tr>
    <th id="e1" headers="header1">1</th>
    <th id="e2" headers="header1">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
  </tr>
</table>
```

### Inapplicable

#### Inapplicable Example 1

There's no usage of the `headers` attribute.

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Exams</th>
  </tr>
  <tr>
    <td>15%</td>
    <td>45%</td>
  </tr>
</table>
```

#### Inapplicable Example 2

A table used for presentation only, that has a `role="presentation"`. 

```html
<table role="presentation">
  <tr>
    <th id="header1">Project Status</th>
  </tr>
  <tr>
    <td>15%</td>
  </tr>
</table>
```

#### Inapplicable Example 3

The rule applies only to `table > td` element.

```html
<div role="table">
  <div role="row">
    <div class="columnheader" id="header1">Projects</div>

>     <div class="columnheader" id="header2">Exams</div>

Reviewer 3: I suppose you meant `role="columnheader"`not `class`

````html

<div role="columnheader" id="header1">Projects</div>
<div role="columnheader" id="header2">Exams</div>

---
id: a25f45
name: headers attribute only refers to cells in the same table element.
rule_type: atomic
description: |
  This rule checks that the headers attribute must only refer to cells in the same table element.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

This rule applies to `td` elements with a `headers` attribute, that are [included in the accessibility tree](#included-in-the-accessibility-tree).

## Expectation

> The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:

Reviewer 4: This seems more like a definition of `header` attribute than an expectation. This is my attempt at improving it:

Each ID in the [set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens) that is the `headers` attribute of the target element, is the ID of an element that:

Author: Sorry Carlos, I feel like what is currently in place is straightforward over your suggestion.

1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and

> 2. is a `cell` of the same [`table`](https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element

Reviewer 4: I don't think cell should be in backticks.
But you might want to link to the definition: [cell](https://www.w3.org/TR/html52/tabular-data.html#cell)

```html
<table>
  <thead>	
    <tr>
      <td role='columnheader' id="header1">Projects</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td headers="header1">15%</td>
    </tr>
  </tbody>
</table>
```

#### Passed Example 4

A `table` with multiple `columnheader(s)`, where the `headers` attribute on the cell lists the `id(s)` of `th` within the same `table`.

```html
<table>
  <tr>
    <th colspan="2" id="header1">Projects</th>
    <th colspan="2" id="header2">Exams</th>
  </tr>
  <tr>
    <th id="e1" headers="header1">1</th>
    <th id="e2" headers="header1">2</th>
    <th id="p1" headers="header2">1</th>
    <th id="p2" headers="header2">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 e2">15%</td>
    <td headers="header2 p1">15%</td>
    <td headers="header2 p2">45%</td>
  </tr>
</table>
```

### Failed

#### Failed Example 1

Cells with `headers` attribute, refers to non-existing `th` within the same `table`.

```html
<table>
  <tr>
    <th id="header1">Projects</th>
    <th id="header2">Exams</th>
  </tr>
  <tr>
    <td headers="header3">15%</td>
    <td headers="header3">15%</td>
  </tr>
</table>
```

> #### Failed Example 2

Author: Updated example 1 to have `NOT-EXIST` as the header id, but keeping fail example 2 as it is slightly different from example 1, given only one of the header id referenced is missing.

> #### Failed Example 2

Reviewer 5: Make the description more of the example more precise:  some headers are OK, 1 is KO and 1 KO is enough.
(this is different from other handlings of list in attributes where the list means "use the first valid one", here the list means "use all of them and they must all be valid")

---
id: a25f45
name: headers attribute only refers to cells in the same table element.
rule_type: atomic
description: |
  This rule checks that the headers attribute must only refer to cells in the same table element.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
input_aspects:
  - DOM Tree
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

This rule applies to `td` elements with a `headers` attribute, that are [included in the accessibility tree](#included-in-the-accessibility-tree).

## Expectation

> Each target element with the `headers` attribute refers to other `th` within the same `table` element.

Author: @carlosapaduarte 
I believe we author most expectations to begin with `Each target element...`.
And the applicability can specify to choose a collection of element.

The language is such that of the many elements that apply to this rule, ensure that each of them satisfy something....

> Each target element with the `headers` attribute refers to other `th` elements or cells with a [semantic role](#semantic-role) of `columnheader` or `rowheader` within the same `table` element.

Reviewer 4: I still think that the text of the Expectation could be improved. What about this:
Each target element refers, in its `headers` attribute, to other `th` elements or cells with a [semantic role](#semantic-role) of `columnheader` or `rowheader` within the same `table` element.

```html
<table>
  <tr>
    <th colspan="2" id="header1">Projects</th>
  </tr>
  <tr>
    <th id="e1" headers="header1">1</th>
    <th id="e2" headers="header1">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
  </tr>
</table>
```

### Inapplicable

#### Inapplicable Example 1

There's no usage of the `headers` attribute.

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Exams</th>
  </tr>
  <tr>
    <td>15%</td>
    <td>45%</td>
  </tr>
</table>
```

#### Inapplicable Example 2

A table used for presentation only, that has a `role="presentation"`. 

```html
<table role="presentation">
  <tr>
    <th id="header1">Project Status</th>
  </tr>
  <tr>
    <td>15%</td>
  </tr>
</table>
```

#### Inapplicable Example 3

The rule applies only to `table > td` element.

```html
<div role="table">
  <div role="row">
    <div class="columnheader" id="header1">Projects</div>
    <div class="columnheader" id="header2">Exams</div>
  </div>
  <div role="row">
    <div role="cell" headers="header2">15%</div>
    <div role="cell" headers="header1">15%</div>
  </div>
</div>
```

#### Inapplicable Example 4

> No `headers` attribute is defined for table `cell`.

Reviewer 4: This is the same as Inapplicable Example 1

---
id: a25f45
name: `headers` attribute only refers to cells in the same table element.
rule_type: atomic
description: |
  This rule checks that the headers attribute must only refer to cells in the same table element.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
input_aspects:
- DOM Tree
authors:
- Jey Nandakumar
- Audrey Maniez
---

## Applicability

> This rule applies to `td` elements with a `headers` attribute, that is [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer 4: is > are

    <th id="p2" headers="header2">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 e2">15%</td>
    <td headers="header2 p1">15%</td>
    <td headers="header2 p2">45%</td>
  </tr>
</table>
```

### Failed

#### Failed Example 1

Cells with `headers` attribute, refers to non-existing `th` within the same `table`.

```html
<table>
  <tr>
    <th id="header1">Projects</th>
    <th id="header2">Exams</th>
  </tr>
  <tr>
    <td headers="header3">15%</td>
    <td headers="header3">15%</td>
  </tr>
</table>
```

#### Failed Example 2

One of the cells with `headers` attribute, refers to `th` that does not exist within the same `table`.

```html
<table>
  <tr>
    <th colspan="2" id="header1">Projects</th>
  </tr>
  <tr>
    <th id="e1" headers="header1">1</th>
    <th id="e2" headers="header1">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
  </tr>
</table>
```

### Inapplicable

#### Inapplicable Example 1

There's no usage of the `headers` attribute.

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Exams</th>
  </tr>
  <tr>
    <td>15%</td>
    <td>45%</td>
  </tr>
</table>
```

#### Inapplicable Example 2

> A table used for presentation only, that has a `role="presentation"`. 

Reviewer 4: Shouldn't the main reason be that there are no `headers` attributes?

```html
<table role="presentation">
  <tr>
    <th colspan="2">My document title</th>
  </tr>
  <tr>
    <td><p>The excerpt</p></td>
    <td><p>The content</p></td>
  </tr>
</table>
```

#### Inapplicable Example 3

> The rule applies only to `table` element.

Reviewer 4: `table` > `td`

```html
<div role="table">
  <div role="row">
    <div class="columnheader" id="header1">Projects</div>
    <div class="columnheader" id="header2">Exams</div>
  </div>
  <div role="row">
    <div role="cell" headers="header2">15%</div>
    <div role="cell" headers="header1">15%</div>
  </div>
</div>
```

#### Inapplicable Example 4

No `headers` attribute is defined for table `cell`.

```html
<table>
  <tr>
    <th id="header1">Projects</th>
    <th id="header2">Exams</th>
  </tr>
  <tr>
    <td colspan="2">15%</td>
  </tr>
</table>
```

#### Inapplicable Example 5

> A table that is not visible. 

Reviewer 4: There is nothing in the applicability about visibility. Not included in the accessibility is covered in the next test case, so you don't need to duplicate.

---
id: a25f45
name: headers attribute only refers to cells in the same table element.
rule_type: atomic
description: |
  This rule checks that the headers attribute must only refer to cells in the same table element.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

This rule applies to `td` elements with a `headers` attribute, that are [included in the accessibility tree](#included-in-the-accessibility-tree).

## Expectation

> The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:

Reviewer 6: Shouldn't it at least be "The value of the `headers` attribute..." ?

Author: we say things like the role attribute elsewhere, so I think this is conveyed by default.

1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and
2. is a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) of the same [`table`](https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element

## Assumptions

- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.
- Tables using `headers` id associations are intended as data rather than [layout table](#layout-table).

> 

Reviewer 6: Should there be another assumption that if the IDs do not live up to the expectations in this rule, the browser will try to calculate a default header from the [algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics), and if the table is a simple table, then header cells might be correctly assigned even though the outcome of this rule is `failed` ?

Author: Adding this will rise more questions like:
- what is do not live up to the expectations?
- what is a `simple table` etc.,

---
id: a25f45
name: Headers attribute of td element refers to cells in the same table element
rule_type: atomic
description: |
  This rule checks that the headers attribute of `td` elements refer to cells in the same `table` element that have a semantic role of `columnheader` or `rowheader`.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

> This rule applies to any `td` element with a `headers` attribute, that is [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer 7: Should the rule not apply to the `headers` attribute instead of the `td` element?

Author: I did some testing on VoiceOver in general `headers` attributes seems to have a lot of inconsistencies. Also tested Firefox + nvda combo, with help of testing team.

Some findings:
- td scope=row does NOT work with VO or NVDA
- using headers="id1 id2" does NOT work 
- using headers to point to row headers in 2nd column does not work in VO, only works in NVDA/Firefox.
- `th` with `headers` attribute were also inconsistent.

Given the inconsistencies, I think it is best to scope this rule to `td` elements for this rule.

Reviewer 7: All good then, though my question still stands: Should the test target be the `headers` attribute instead of the `td` element? This would be consistent with other rules that solely target attributes.

## Expectation

> The `header` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:

Reviewer 7: The expectation needs to account for the presence of nested browsing contexts and Shadow DOM:

> ...each of which is an ID of an element **in the same document tree or shadow tree as the applicable element**.

---
id: a25f45

> name: Headers attribute of td element refers to cells in the same table element

Reviewer: The applicability uses "cell" as a definition, rather than `td`. A cell, according to spec, can be either a `td` or a `th`, so the applicability is inconsistent with the rule name and description. In my opinion this rule shouldn't be limited to `td` elements, so I think the applicability is correct, but the name / description aren't, and there are missing test cases for `th` with a `headers` attribute.

rule_type: atomic
description: |

>   This rule checks that the headers attribute of `td` elements refer to cells in the same `table` element that has a semantic role of `columnheader` or `rowheader`.

Reviewer 4: cells ... that has a semantic role > cells ... that have a semantic role

accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

> This rule applies to any `td` element with a `headers` attribute, that are [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer 4: that are > that is

> This rule applies to any [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) with a [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute, in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer 7: Still not addressed: https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r311592701. It could look like this:

> This rule applies to any [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).

Author: Good play in words. Will switcheroo.

> This rule applies to any [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer: I don't think you should link the headers attribute or the table element. We're not doing that with element / attribute names in other rules. Also, MDN links don't include definitions, and those pages aren't stable.

> This rule applies to any [`headers`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-headers) attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a [table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table) element that is [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer: This rule probably shouldn't be applicable to `table` elements with a semantic role of `table` or `grid`. If the semantic role of table is changed, the heading assignment no longer applies. I'm not sure if this happens consistently in AT, so that might be worth testing.

## Expectation

> The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:

Reviewer: You don't need to look at trees here. A cell in the same table (the second bullet) can't be in a different tree. This part should simply be:

> The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element that:

Reviewer 7: Allow me to abuse the DOM yet again:

```html
<table>
  <th>
    <div></div>
  </th>
  <td headers="foo"></td>
</table>

<script>
  const th = document.createElement("th");
  th.id = "foo";

  document
    .querySelector("div")
    .attachShadow({ mode: "open" })
    .appendChild(th);
</script>
```

Author: @WilcoFiers seeing this comment - https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r318053483, it think it is best to emphasise document/ shadow tree. Wdyt?

Reviewer: @kasperisager that `th` element isn't in the same tree as the td (see the definitions of tree)

Reviewer 7: @WilcoFiers Bah, I copy pasted the wrong code:

```html
<table>
  <th>
    <div></div>
  </th>
  <td headers="foo"></td>
</table>

<script>
  const th = document.createElement("th");
  th.id = "foo";

  document
    .querySelector("div")
    .appendChild(th);
</script>
```

Author: Leaving usages of `document tree` and `shadow tree`.

> The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:

Reviewer 5: Editorial: elment => element (end of the sentence)

> The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://www.w3.org/TR/dom41/#document-trees) or [shadow tree](https://www.w3.org/TR/dom41/#shadow-trees) as the applicable elment, that:

Reviewer 5: What happen if there are several element with that id?
```html
<table>
  <thead>	
    <tr>
      <th id="header1">Projects</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td headers="header1">15%</td>
    </tr>
<tr>
      <td id="header1">25%</td>
</tr>
  </tbody>
</table>
```

Author: This could be more developer error than an actual scenario. I did VoiceOver testing and in some cases one is announced or none :/. I am unsure if this is within the scope of this rule.

Reviewer 5: I agree that this would be more of a global HTML error than a problem with this rule.
Maybe it's worth an assumption "This rule assume that `id` are unique within the current document" to ensure that we can write "the ID of **an** element" without any ambiguity.

> 1. has a [semantic role](#semantic-role) of `columnheader` or `rowheader`, and

Reviewer: I don't really understand why this is required. I know the spec says that you should only reference other headers with the `headers` attribute, but are there any accessibility issues that come from referencing a `td` element with the `headers` attribute? I don't see what they would be, unless maybe assistive tech don't announce `td`s if they're used in the `headers` attribute, but I don't think that's the case.

Author: Note to self:
- Get data of AT testing.

Author: @WilcoFiers 

I did some testing & other `td` referred via `headers` attribute is not consistently announced without a role of `columnheder` or `rowheader`. These tests were done with voiceover.

Note voiceover seems to be very inconsistent even when `th` is referred in `headers` attribute. It also seems like headers are only rightly announced when wrapped with `<thead>` in the case of voiceover.

Here is also an article on elaborate testing outcomes:
- [VoiceOver and Tables](http://accessibleculture.org/articles/2010/10/voiceover-and-tables-with-an-empty-first-header-cell/)

I did not test all of the test cases listed in the above example.

Note though in terms of automated implementation of this rule, I do think it does not matter whether a header attribute refers to `th` or `td`.

Author: Did more testing referencing `td` with `headers` attribute, does not get announced in VoiceOver :/. Overall headers attribute is quite inconsistent with AT's.

2. is a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) of the same [`table`](https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element

## Assumptions

- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.

> - This test assumes that if the IDs of the `headers` attribute, do not live up to the expectations in this rule, the browser will try to calculate a default header from the [algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics).

Reviewer: I think it's actually the opposite. We're assuming that if `headers` attribute fails, the browser won't be able to come up with the correct headers. Also, use "expectation" rather than "expectations", since there's only one of them.

```html
<table>
  <thead>	
    <tr>
      <td role='columnheader' id="header1">Projects</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td headers="header1">15%</td>
    </tr>
  </tbody>
</table>
```

#### Passed Example 4

A `table` with multiple `columnheader(s)`, where the `headers` attribute on the cell lists the `id(s)` of `th` within the same `table`.

```html
<table>
  <tr>
    <th colspan="2" id="header1">Projects</th>
    <th colspan="2" id="header2">Exams</th>
  </tr>
  <tr>
    <th id="e1" headers="header1">1</th>
    <th id="e2" headers="header1">2</th>
    <th id="p1" headers="header2">1</th>
    <th id="p2" headers="header2">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 e2">15%</td>
    <td headers="header2 p1">15%</td>
    <td headers="header2 p2">45%</td>
  </tr>
</table>
```

#### Passed Example 5

The `headers` attribute on the cell refers to `th` with a role of `rowheader` within the same `table`.

```html
<table>
  <tbody>
    <tr>
      <th role="rowheader" id="headerAge">Age</th>
      <td headers="headerAge">65</td>
    </tr>
  </tbody>
</table>
```

> ### Failed

Reviewer: Out of all of these, I think only failed example 3 actually fails SC 1.3.1. For the sake of having all failed examples be true failures of the SC, I think these things are going to need some work.

> ### Failed

Reviewer: There is no test for when `headers` refers to a data call. I don't think that should fail this rule, so if that is changed this comment is resolved as well.

Author: This feels like something an implementation should consider to return an outcome for these test cases, over a change in the rule?

    <th id="e2" headers="header1">2</th>
  </tr>
  <tr>
    <td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
  </tr>
</table>
```

#### Failed Example 3

The `headers` attribute on the cell refers to an element outside the same `table`.

```html
<span id="elmOutsideTable">Project Costs</span>

<table>
  <tr>
    <th id="headerOfColumn">Projects</th>
  </tr>
  <tr>
    <td headers="elmOutsideTable">15%</td>
  </tr>
</table>
```

#### Failed Example 4

The `headers` attribute on the cell refers to an element inside the same `table`, but is not a `rowheader` or `columnheader`.

```html
<table>
  <tr>
     <td>
      <span id="headerProject">Projects</span>
    </td>
  </tr>
  <tr>
    <td headers="headerProject">
      15% 
    </td>
  </tr>
</table>
```

### Inapplicable

#### Inapplicable Example 1

There's no usage of the `headers` attribute.

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Exams</th>
  </tr>
  <tr>
    <td>15%</td>
    <td>45%</td>
  </tr>
</table>
```

#### Inapplicable Example 2

A table used for presentation only, that has a `role="presentation"`. 

```html
<table role="presentation">
  <tr>

>     <th id="header1">Project Status</th>

Reviewer: Putting a `th` in a layout table may be controversial. I don't think it fails WCAG, but I know not everyone agrees with that. Putting this example in the way it is now means this rule is testing something that it's not about. I'd suggest changing the `th` to a `td` here.

  </tr>
  <tr>
    <td>15%</td>
  </tr>
</table>
```

#### Inapplicable Example 3

The rule applies only to `table > td` element.

```html
<div role="table">
  <div role="row">
    <div role="columnheader" id="header1">Projects</div>
    <div role="columnheader" id="header2">Exams</div>
  </div>
  <div role="row">
    <div role="cell" headers="header2">15%</div>
    <div role="cell" headers="header1">15%</div>
  </div>
</div>
```

#### Inapplicable Example 4

> A table that is not included in the accessibility tree. 

Reviewer: This would be inapplicable even without aria-hidden. Please add the `headers` attribute.

---
id: a25f45
name: Headers attribute specified on a cell refers to cells in the same table element
rule_type: atomic
description: |
  This rule checks that the headers attribute on a cell refer to other cells in the same `table` element with a semantic role of `columnheader` or `rowheader`.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Jey Nandakumar
    - Audrey Maniez
---

## Applicability

> This rule applies to any `headers` attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) within a `table` element, where the `table` element is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer: I don't think the visibility of the table actually matters here. The headers attribute isn't visible, it's only available to assistive tech, why would a table need to be visible for this rule to apply?

Reviewer 2: Since 1.3.1 cares about the programmatic associations matching visually conveyed associations, I think the visibility matters.

Reviewer: Emma's convinced me, let's leave this as is.

## Expectation

Each target attribute is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is the ID of an element, that:

> 1. has a [semantic role](#semantic-role) of [columnheader](https://www.w3.org/TR/html-aria/#index-aria-columnheader) or [rowheader](https://www.w3.org/TR/html-aria/#index-aria-rowheader), and

Reviewer 2: The rule is not specific about what the element is or isn't, only that it has an expected semantic role and is part of the same table. 

As an atomic rule checking a particular thing, a fail scenario would be inaccessible, while a pass scenario would be part of a wider set of checks for a SC. So I'm not seeing how the comment is still relevant.

> 1. has a [semantic role](#semantic-role) of [columnheader](https://www.w3.org/TR/html-aria/#index-aria-columnheader) or [rowheader](https://www.w3.org/TR/html-aria/#index-aria-rowheader), and

Reviewer 8: That would be a problem. Wilco states that this should not be required because it will fail situations that aren't actually a WCAG issue (VoiceOver not reading the headers correctly would not constitute a WCAG failure per se).

2. is a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) of the same [`table`](https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element.

## Assumptions

- This rule assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for scripts, this rule may not be accurate.
- This rule assumes that the `headers` attribute is required to express the relationship between data and header cells in the `table`. If the browser [computes an adequate fallback header](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics), this rule may produce incorrect results.

## Accessibility Support

> - `headers` attribute may not be consistently announced by assistive technologies.

Reviewer: This is too much of a generalisation?

```suggestion
The `headers` attribute is not supported by some popular assistive technologies as a way to assign headers in a table. Passing this rule can still cause issues for users of those assistive technologies.
```

		</tr>
	</tbody>
</table>
```

#### Passed Example 4

This `table` has multiple elements with a role of `columnheader`. The `headers` attribute on the cells lists IDs of `th` elements within the same `table`.

```html
<table>
	<tr>
		<th colspan="2" id="header1">Projects</th>
		<th colspan="2" id="header2">Exams</th>
	</tr>
	<tr>
		<th id="e1" headers="header1">1</th>
		<th id="e2" headers="header1">2</th>
		<th id="p1" headers="header2">1</th>
		<th id="p2" headers="header2">2</th>
	</tr>
	<tr>
		<td colspan="2" headers="header1 e1 e2">15%</td>
		<td headers="header2 p1">15%</td>
		<td headers="header2 p2">45%</td>
	</tr>
</table>
```

#### Passed Example 5

The `headers` attribute on the cell refers to a `th` element with a role of `rowheader` within the same `table`.

```html
<table>
	<tbody>
		<tr>
			<th role="rowheader" id="headerAge">Age</th>
			<td headers="headerAge">65</td>
		</tr>
	</tbody>
</table>
```

#### Passed Example 6

The `headers` attribute on the last two `th` elements refers to another `th` element within the same `table`. Here the column header has a span of two columns.

```html
<table>
	<tr>
		<th id="name" colspan="2">Name</th>
	</tr>
	<tr>

> 		<th headers="name">Firsname</th>

Reviewer 2: Minor typo in the cell content

		<th headers="name">Lastname</th>
	</tr>
</table>
```

### Failed

#### Failed Example 1

The `headers` attribute on the cell refers to an element outside the same `table`.

```html
<span id="elmOutsideTable">Project Costs</span>

<table>
	<tr>
		<th id="headerOfColumn">Projects</th>
	</tr>
	<tr>

> 		<td headers="elmOutsideTable">15%</td>

Reviewer: This comment was never addressed: https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r318026911 Like I said before, I don't think failed example 1 and 2 actually fail the success criterion.

Reviewer 2: I disagree, sort of. Reading 1.3.1 is seems to be about the programatic information conveying the same association that is conveyed visually.

Failed examples 1 and 2 certainly match the rule expectations.

For failed example 1, it could be more clear that the span does not visually head up the table cell. For failed example 2, it could be that the Id doesn't exist in the page at all.

It strikes me that both examples are maybe an `aria-labelledby` away from not being a fail example. A point of view that raises questions about the rule and the referenced SC technique possibly not being detailed enough.

Reviewer: This needs a little research.

Reviewer 5: As per the [Algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#algorithm-for-assigning-header-cells), the position of the cell is only looked at if there is no `headers` attribute.

So, I think that in this case there is no programmatic association. (maybe UAs and ATs are fallbacking to the position if the `headers` attribute is KO...)

Reviewer: Every screen reader we've tested this in will look at row/column headers for fallback. That may not be what the algorithm says they should do, but someone testing with one of the top 5 AT / browser combinations is not going to see a problem here. I find this test case questionable. Here's the test case we ran a little while ago:

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Progress</th>
  </tr>
  <tr>
    <td headers="projects100">My Project</td>
    <td>15%</td>
  </tr>
</table>
```
**IE + JAWS**: Works
**Chrome + JAWS**: Works
**Firefox + NVDA**: Works
**Safari + VO**: Works

Reviewer 5: If this passes the SC (which I don't challenge), then the rule need to be changed to take fallback headers into account.
This certainly fails the rule right now,  so if this passes the SC, the rule is bad.

And the rule in its current version is not referencing the assignation algorithm anymore…

(technically, the rule does assume that there is no fallback headers, but if this assumption is false for the vaste majority of cases, it needs to go away)

	</tr>
</table>
```

#### Failed Example 2

The `td` element has a `headers` attribute referring to an ID that does not exist within the same `table` (or within the [document](https://dom.spec.whatwg.org/#concept-document)). Even if some of the ID referred to exist, a single non-existing one is enough to fail the rule.

```html
<table>
	<tr>
		<th colspan="2" id="header1">Projects</th>
	</tr>
	<tr>
		<th id="e1" headers="header1">1</th>
		<th id="e2" headers="header1">2</th>
	</tr>
	<tr>
		<td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
	</tr>
</table>
```

#### Failed Example 3

The `headers` attribute on the cell refers to an element inside the same `table` which does not have a role of `rowheader` or `columnheader`.

```html
<table>
	<tr>
		<td>
			<span id="headerProject">Projects</span>
		</td>
	</tr>
	<tr>
		<td headers="headerProject">
			15%
		</td>
	</tr>
</table>
```

> ### Inapplicable

Reviewer: You didn't test "not included in the accessibility tree".

#### Inapplicable Example 1

There is no `headers` attribute.

```html
<table>
	<tr>
		<th scope="col">Projects</th>
		<th scope="col">Exams</th>
	</tr>
	<tr>
		<td>15%</td>
		<td>45%</td>
	</tr>
</table>
```

#### Inapplicable Example 2

The `table` has a `role="presentation"` and thus is not [included in the accessibility tree](#included-in-the-accessibility-tree).

```html
<table role="presentation">
	<tr>
		<td id="header1">Project Status</td>
	</tr>
	<tr>
		<td>15%</td>
	</tr>
</table>
```

#### Inapplicable Example 3

The `table` is not [visible](#visible) in page.

```html
<html>
	<style>
		.notInPage {
			position: absolute;
			left: -9999px;
			top: -9999px;
		}
	</style>
	<table class="notInPage">
		<tr>
			<td id="header1">Project Status</td>
		</tr>
		<tr>

> 			<td>15%</td>

Reviewer 2: Should this have a headers attribute? Or should the "header1" be a `<th>` with no id?

		</tr>
	</table>
</html>
```

#### Inapplicable Example 4

The rule applies only to `headers` attribute within a `table` element.

```html
<div role="table">
	<div role="row">
		<div role="columnheader" id="header1">Projects</div>
		<div role="columnheader" id="header2">Exams</div>
	</div>
	<div role="row">
		<div role="cell" headers="header2">15%</div>
		<div role="cell" headers="header1">15%</div>
	</div>
</div>

> ```

Reviewer 2: Will a different rule check this kind of thing? This code could be styled to look just like a table, and has all the correct attributes.

Reviewer: Yes, it will be tested by a different rule. The headers attribute does not work on div. It can only be on td / th. This should be in the description.

---
id: a25f45
name: Headers attribute specified on a cell refers to cells in the same table element
rule_type: atomic
description: |
  This rule checks that the headers attribute on a cell refer to other cells in the same `table` element with a semantic role of `columnheader` or `rowheader`.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
authors:
  - Jey Nandakumar
  - Audrey Maniez
---

## Applicability

> This rule applies to any `headers` attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) in a `table` element that is is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree).

Author: I am unsure about hidden cells within a table & the scope of this rule to cater to them. Can other reviewers voice their opinions?

## Expectation

The `headers` attribute of each target element is [a set of space separated IDs](https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens), each of which is an ID of an element in the same [document tree](https://dom.spec.whatwg.org/#document-trees) or [shadow tree](https://dom.spec.whatwg.org/#shadow-trees) as the applicable element, that:

1. has a [semantic role](#semantic-role) of [columnheader](https://www.w3.org/TR/html-aria/#index-aria-columnheader) or [rowheader](https://www.w3.org/TR/html-aria/#index-aria-rowheader), and
2. is a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) of the same [`table`](https://www.w3.org/TR/html50/tabular-data.html#concept-table) as the target element.

## Assumptions

- This test assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for script, this rule may not be accurate.

> - This test assumes that the `headers` attribute is required to express the relationship between data and header cells in the `table`. If the browser [computes an adequate fallback header](https://html.spec.whatwg.org/multipage/tables.html#header-and-data-cell-semantics), this rule may produce incorrect results.

Reviewer 8: The assumption is there because sometimes the relationship is expressed in different ways or sometimes there is no relationship at all and the header attribute is added in error (but also has a non existing ID). This assumption covers those situations.

```html
<table>
	<tbody>
		<tr>
			<th role="rowheader" id="headerAge">Age</th>
			<td headers="headerAge">65</td>
		</tr>
	</tbody>
</table>
```

#### Passed Example 6

The `headers` attribute on the `th` cell refers to other `th` cells within the same `table`. Here the column header has a spans two columns.

```html
<table>
	<tr>
		<th id="name" colspan="2">Name</th>
	</tr>
	<tr>
		<th headers="name">Firsname</th>
		<th headers="name">Lastname</th>
	</tr>
</table>
```

### Failed

#### Failed Example 1

The `headers` attribute on the cell refers to an element outside the same `table`.

```html
<span id="elmOutsideTable">Project Costs</span>

<table>
	<tr>
		<th id="headerOfColumn">Projects</th>
	</tr>
	<tr>
		<td headers="elmOutsideTable">15%</td>
	</tr>
</table>
```

#### Failed Example 2

One of the cells with `headers` attribute, refers to `th` that does not exist within the same `table`.

```html
<table>
	<tr>
		<th colspan="2" id="header1">Projects</th>
	</tr>
	<tr>
		<th id="e1" headers="header1">1</th>
		<th id="e2" headers="header1">2</th>
	</tr>
	<tr>
		<td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
	</tr>
</table>
```

#### Failed Example 3

> Cells with `headers` attribute, refers to multiple non-existing elements within the same `table`.

Reviewer 5: ```suggestion
Both `td` elements have `headers` attribute which refers to ID that do not exist within the same `table`.
```

```html
<table>
	<tr>
		<th id="header1">Projects</th>
		<th id="header2">Exams</th>
	</tr>
	<tr>
		<td headers="NOT-EXIST-1">15%</td>
		<td headers="NOT-EXIST-2">15%</td>
	</tr>
</table>
```

#### Failed Example 4

> The `headers` attribute on the cell refers to an element inside the same `table`, but is not a `rowheader` or `columnheader`.

Reviewer 5: ```suggestion
The `headers` attribute on the cell refers to an element inside the same `table` which does not have a role of `rowheader` or `columnheader`.
```

```html
<table>
	<tr>
		<td>
			<span id="headerProject">Projects</span>
		</td>
	</tr>
	<tr>
		<td headers="headerProject">
			15%
		</td>
	</tr>
</table>
```

> ### Inapplicable

Reviewer 5: Need an example with a `table` that is not visible (but included in the accessibility tree).

#### Inapplicable Example 1

> There's no usage of the `headers` attribute.

Reviewer 5: ```suggestion
There is no `headers` attribute.
```

```html
<table>
	<tr>
		<th scope="col">Projects</th>
		<th scope="col">Exams</th>
	</tr>
	<tr>
		<td>15%</td>
		<td>45%</td>
	</tr>
</table>
```

#### Inapplicable Example 2

> A table used for presentation only, that has a `role="presentation"`.

Reviewer 5: ```suggestion
The `table` has a `role="presentation"` and thus is not [included in the accessibility tree](#included-in-the-accessibility-tree).
```

```html
<table role="presentation">
	<tr>
		<td id="header1">Project Status</td>
	</tr>
	<tr>
		<td>15%</td>
	</tr>
</table>
```

#### Inapplicable Example 3

> The rule applies only to `table` element.

Reviewer 5: ```suggestion
The rule applies only to `headers` attribute within a `table` element.
```

```html
<div role="table">
	<div role="row">
		<div role="columnheader" id="header1">Projects</div>
		<div role="columnheader" id="header2">Exams</div>
	</div>
	<div role="row">
		<div role="cell" headers="header2">15%</div>
		<div role="cell" headers="header1">15%</div>
	</div>
</div>
```

#### Inapplicable Example 4

> A table that is not included in the accessibility tree.

Reviewer 5: ```suggestion
The `table` is not [included in the accessibility tree](#included-in-the-accessibility-tree).
```

```html
<table>
	<tr>
		<th scope="col">Projects</th>
		<th scope="col">Exams</th>
	</tr>
	<tr>
		<td>15%</td>
		<td>45%</td>
	</tr>
</table>
```

#### Inapplicable Example 2

The `table` has a `role="presentation"` and thus is [decorative](#decorative).

```html
<table role="presentation">
	<tr>
		<td id="header1">Project Status</td>
	</tr>
	<tr>
		<td headers="header1">15%</td>
	</tr>
</table>
```

#### Inapplicable Example 3

The `table` is not [visible](#visible) in page.

```html
<html>
	<style>
		.notInPage {
			position: absolute;
			left: -9999px;
			top: -9999px;
		}
	</style>
	<table class="notInPage">
		<tr>
			<td id="header1">Project Status</td>
		</tr>
		<tr>
			<td headers="header1">15%</td>
		</tr>
	</table>
</html>
```

#### Inapplicable Example 4

The rule applies only to `headers` attribute within a `table` element.

```html
<div role="table">
	<div role="row">
		<div role="columnheader" id="header1">Projects</div>
		<div role="columnheader" id="header2">Exams</div>
	</div>
	<div role="row">
		<div role="cell" headers="header2">15%</div>
		<div role="cell" headers="header1">15%</div>
	</div>
</div>
```

#### Inapplicable Example 5

> The `table` is not [included in the accessibility tree][].

Reviewer 8: faulty brackets around link

---
id: a25f45
name: Headers attribute specified on a cell refers to cells in the same table element
rule_type: atomic
description: |
  This rule checks that the headers attribute on a cell refer to other cells in the same `table` element with a semantic role of `columnheader` or `rowheader`.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Jey Nandakumar
    - Audrey Maniez
---

## Applicability

> This rule applies to any `headers` attribute specified on a [`cell`](https://www.w3.org/TR/html52/tabular-data.html#cell) within a `table` element that is is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree).

Reviewer 5: Even without considering invisible cells, you can still rephrase that to make it obvious that the "that is visible and included in the accessibility tree" part refers to the `table`, not the `cell`.

## Test Cases

### Passed

#### Passed Example 1

The `headers` attribute on the cell refers to a `th` element within the same `table`.

```html
<table>
	<thead>
		<tr>
			<th id="header1">Projects</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td headers="header1">15%</td>
		</tr>
	</tbody>
</table>
```

#### Passed Example 2

The `headers` attribute on the cell refers to a `th` element within the same `table`. Multiple headers are referenced for a cell with `colspan` of `2`.

```html
<table>
	<thead>
		<tr>
			<th id="header1">Projects</th>
			<th id="header2">Exams</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td colspan="2" headers="header1 header2">15%</td>
		</tr>
	</tbody>
</table>
```

#### Passed Example 3

The `headers` attribute on the second cell refers to a `td` element with a role of `columnheader` within the same `table`.

```html
<table>
	<thead>
		<tr>
			<td role="columnheader" id="header1">Projects</td>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td headers="header1">15%</td>
		</tr>
	</tbody>
</table>
```

#### Passed Example 4

> This `table` has multiple `columnheader(s)`. The `headers` attribute on the cells lists the `id(s)` of `th` elements within the same `table`.

Reviewer 5: ```suggestion
This `table` has multiple elements with a role of `columnheader`. The `headers` attribute on the cells lists IDs of `th` elements within the same `table`.
```

The `headers` attribute on the last two `th` elements refers to another `th` element within the same `table`. Here the column header has a span of two columns.

```html
<table>
	<tr>
		<th id="name" colspan="2">Name</th>
	</tr>
	<tr>
		<th headers="name">Firsname</th>
		<th headers="name">Lastname</th>
	</tr>
</table>
```

### Failed

#### Failed Example 1

The `headers` attribute on the cell refers to an element outside the same `table`.

```html
<span id="elmOutsideTable">Project Costs</span>

<table>
	<tr>
		<th id="headerOfColumn">Projects</th>
	</tr>
	<tr>
		<td headers="elmOutsideTable">15%</td>
	</tr>
</table>
```

#### Failed Example 2

The `td` element has a `headers` attribute referring to an ID that does not exist within the same `table` (or within the [document](https://dom.spec.whatwg.org/#concept-document)). Even if some of the ID referred to exist, a single non-existing one is enough to fail the rule.

```html
<table>
	<tr>
		<th colspan="2" id="header1">Projects</th>
	</tr>
	<tr>
		<th id="e1" headers="header1">1</th>
		<th id="e2" headers="header1">2</th>
	</tr>
	<tr>
		<td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>
	</tr>
</table>
```

> #### Failed Example 3

Reviewer 5: This is quite redundant with Failed Example 2, but simplified. Could be removed.

Both `td` elements have `headers` attribute which refers to ID that do not exist within the same `table`.

```html
<table>
	<tr>
		<th id="header1">Projects</th>
		<th id="header2">Exams</th>
	</tr>
	<tr>
		<td headers="NOT-EXIST-1">15%</td>
		<td headers="NOT-EXIST-2">15%</td>
	</tr>
</table>
```

#### Failed Example 4

The `headers` attribute on the cell refers to an element inside the same `table` which does not have a role of `rowheader` or `columnheader`.

```html
<table>
	<tr>
		<td>
			<span id="headerProject">Projects</span>
		</td>
	</tr>
	<tr>
		<td headers="headerProject">
			15%
		</td>
	</tr>
</table>
```

### Inapplicable

#### Inapplicable Example 1

There is no `headers` attribute.

```html
<table>
	<tr>
		<th scope="col">Projects</th>
		<th scope="col">Exams</th>
	</tr>
	<tr>
		<td>15%</td>
		<td>45%</td>
	</tr>
</table>
```

#### Inapplicable Example 2

The `table` has a `role="presentation"` and thus is not [included in the accessibility tree](#included-in-the-accessibility-tree).

```html
<table role="presentation">
	<tr>
		<td id="header1">Project Status</td>
	</tr>
	<tr>
		<td>15%</td>
	</tr>
</table>
```

#### Inapplicable Example 3

> The `table` is not [visible](#visible).

Reviewer 5: Rather put the table off-screen to make it not visible but still included in the accessibility tree.

```html
<table role="presentation">
	<tr>
		<td id="header1">Project Status</td>
	</tr>
	<tr>
		<td>15%</td>
	</tr>
</table>

> ```

Reviewer 5: This is the same example as inapplicable example 2 where the table was **not** included in the accessibility tree…

#### Inapplicable Example 4

The rule applies only to `headers` attribute within a `table` element.

```html
<div role="table">
	<div role="row">
		<div role="columnheader" id="header1">Projects</div>
		<div role="columnheader" id="header2">Exams</div>
	</div>
	<div role="row">
		<div role="cell" headers="header2">15%</div>
		<div role="cell" headers="header1">15%</div>
	</div>
</div>
```

> #### Inapplicable Example 5

Reviewer 5: Essentially the same as Inapplicable example 2, and we're not going to list all the ways a table can be excluded from the accessibility tree. Can be removed.

---
id: a25f45
name: Headers attribute specified on a cell refers to cells in the same table element
rule_type: atomic
description: |
  This rule checks that the headers attribute on a cell refer to other cells in the same `table` element.
accessibility_requirements:
  wcag20:1.3.1: # Info and Relationships (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS Styling
acknowledgements:
  authors:
    - Jey Nandakumar
    - Audrey Maniez
---

## Applicability

This rule applies to any `headers` attribute specified on a [`cell`][] within a [`table`][] element, where the [`table`][] element is [visible][] and [included in the accessibility tree][].

## Expectation

Each target attribute is [a set of space separated IDs][], each of which is the ID of an element, that:

> 1. is a [`cell`][] of the same [`table`][] as the target element.

Reviewer: Why is this a list of 1? I don't think this should be a list if it has only 1 item.

## Assumptions

- This rule assumes that the `headers` attribute is only used to identify table headers. If other information is included in the `headers` attribute, the rule may fail on issues that are not accessibility concerns. For example, if `headers` is used to include information for scripts, this rule may not be accurate.

> - This rule assumes that the `headers` attribute is required to express the relationship between data and header cells in the `table`. If the browser [computes an adequate fallback header][], this rule may produce incorrect results.

Reviewer 5: If this assumption is false for most cases (https://github.com/act-rules/act-rules.github.io/pull/412/files?file-filters%5B%5D=.md#r378811922) it needs to go away, and the rule updated.

## Accessibility Support

> The `headers` attribute is not supported by some popular assistive technologies as a way to assign headers in a table. Passing this rule can still cause issues for users of those assistive technologies.

Reviewer: The test data @jkodu gave me a little while ago showed that the `headers` attribute works in all major AT. Can you name me a popular browser / AT combination that doesn't work with the `headers` attribute?

Author: Agreed. For reference here is the test data - https://github.com/act-rules/act-rules.github.io/wiki/Assistive-technology-tests-on-table-header-associations

	</tr>
	<tr>
		<td headers="projects1">My Project</td>
		<td headers="progress1">15%</td>
	</tr>
</table>
```

#### Passed Example 8

The `headers` attribute on the cell refers to `th` element which is not the same column as the cell.

```html
<table>
	<tr>
		<td></td>
		<th id="projects2">Projects</th>
	</tr>
	<tr>
		<td headers="projects2">15%</td>
		<td></td>
	</tr>
</table>
```

### Failed

#### Failed Example 1

The `headers` attribute on the cell refers to an element outside the same `table`.

```html
<span id="elmOutsideTable">Project Costs</span>
<table>
	<tr>
		<th id="headerOfColumn">Projects</th>
	</tr>
	<tr>
		<td headers="elmOutsideTable">15%</td>
	</tr>
</table>
```

#### Failed Example 2

The `td` element has a `headers` attribute referring to an ID that does not exist within the same `table` (or within the [document](https://dom.spec.whatwg.org/#concept-document)). Even if some of the ID referred to exist, a single non-existing one is enough to fail the rule.

```html
<table>
	<tr>
		<th colspan="2" id="header1">Projects</th>
	</tr>
	<tr>
		<th id="e1" headers="header1">1</th>
		<th id="e2" headers="header1">2</th>
	</tr>
	<tr>

> 		<td colspan="2" headers="header1 e1 NOT-EXIST">15%</td>

Reviewer: Just like the previous example, this table is perfectly accessible. The first assumption is true for this test case. This example needs to actually show a case where the assumption isn't true, for example, mistype the value in the `headers` attribute on the `th`.

Reviewer 5: I don't think we should have examples where some assumptions are **false** (these would be examples where the rule cannot work, so its result is not really important).

Again, this does fail the rule in its current state. If this passes the SC (again, not challenging the fact), the rule needs to be changed.

The `table` is not [visible][] in page.

```html
<html>
	<style>
		.notInPage {
			position: absolute;
			left: -9999px;
			top: -9999px;
		}
	</style>
	<table class="notInPage">
		<tr>
			<td id="header1">Project Status</td>
		</tr>
		<tr>
			<td headers="header1">15%</td>
		</tr>
	</table>
</html>
```

#### Inapplicable Example 4

The rule applies only to `headers` attribute within a `table` element.

```html
<div role="table">
	<div role="row">
		<div role="columnheader" id="header1">Projects</div>
		<div role="columnheader" id="header2">Exams</div>
	</div>
	<div role="row">
		<div role="cell" headers="header2">15%</div>
		<div role="cell" headers="header1">15%</div>
	</div>
</div>
```

#### Inapplicable Example 5

The `table` is not [included in the accessibility tree][].

```html
<table style="display:none;">
	<tr>
		<td id="header1">Project Status</td>
	</tr>
	<tr>
		<td headers="header1">15%</td>
	</tr>
</table>
```

[included in the accessibility tree]: #included-in-the-accessibility-tree 'Definition of included in the accessibility tree'
[visible]: #visible 'Definition of visible'
[decorative]: #marked-as-decorative 'Definition of marked as decorative'

> [`cell`]: https://www.w3.org/TR/html52/tabular-data.html#cell 'Definition of cell'

Reviewer: Use whatwg here

[semantic role]: #semantic-role 'Definition of semantic role'

> [a set of space separated ids]: https://www.w3.org/TR/html50/infrastructure.html#set-of-space-separated-tokens 'Space separated tokens'

Reviewer: why is this HTML 5.0? Use whatwg

> [`table`]: https://www.w3.org/TR/html50/tabular-data.html#concept-table 'Definition of table'

Reviewer: As above
